File: lib/screens/projects_library_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../providers/library_provider.dart';
import '../providers/project_provider.dart';

class ProjectsLibraryScreen extends StatefulWidget {
  const ProjectsLibraryScreen({super.key});

  @override
  State<ProjectsLibraryScreen> createState() => _ProjectsLibraryScreenState();
}

class _ProjectsLibraryScreenState extends State<ProjectsLibraryScreen> {
  final TextEditingController _searchController = TextEditingController();
  String _searchQuery = '';

  @override
  void initState() {
    super.initState();
    _searchController.addListener(() {
      setState(() {
        _searchQuery = _searchController.text.toLowerCase();
      });
    });
  }

  @override
  void dispose() {
    _searchController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final libraryProvider = Provider.of<LibraryProvider>(context);
    final projectProvider = Provider.of<ProjectProvider>(context, listen: false);

    final filteredProjects = libraryProvider.projects.where((project) {
      final matchesName = project.name.toLowerCase().contains(_searchQuery);
      final matchesDesc = project.description.toLowerCase().contains(_searchQuery);
      final matchesTags = project.tags.any((tag) => tag.toLowerCase().contains(_searchQuery));
      return matchesName || matchesDesc || matchesTags;
    }).toList();

    return Scaffold(
      appBar: AppBar(
        title: const Text('My Projects Library'),
      ),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(16.0),
            child: TextField(
              controller: _searchController,
              decoration: InputDecoration(
                labelText: 'Search projects (name, description, tags)',
                prefixIcon: const Icon(Icons.search),
                suffixIcon: _searchQuery.isNotEmpty
                    ? IconButton(
                        icon: const Icon(Icons.clear),
                        onPressed: () {
                          _searchController.clear();
                        },
                      )
                    : null,
                border: const OutlineInputBorder(),
              ),
            ),
          ),
          Expanded(
            child: filteredProjects.isEmpty
                ? const Center(child: Text('No projects found.'))
                : ListView.builder(
                    itemCount: filteredProjects.length,
                    itemBuilder: (context, index) {
                      final project = filteredProjects[index];
                      return Card(
                        margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
                        child: ListTile(
                          title: Text(project.name, style: const TextStyle(fontWeight: FontWeight.bold)),
                          subtitle: Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              if (project.description.isNotEmpty)
                                Text(project.description, maxLines: 2, overflow: TextOverflow.ellipsis),
                              const SizedBox(height: 4),
                              Wrap(
                                spacing: 4,
                                children: project.tags.map((tag) => ActionChip(
                                  label: Text(tag, style: const TextStyle(fontSize: 10)),
                                  padding: EdgeInsets.zero,
                                  visualDensity: VisualDensity.compact,
                                  onPressed: () {
                                    _searchController.text = tag;
                                  },
                                )).toList(),
                              ),
                              Text(
                                'Last modified: ${project.lastModified.toString().split('.')[0]}',
                                style: const TextStyle(fontSize: 10, color: Colors.grey),
                              ),
                            ],
                          ),
                          trailing: Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              IconButton(
                                icon: const Icon(Icons.edit),
                                tooltip: 'Load Project',
                                onPressed: () {
                                  showDialog(
                                    context: context,
                                    builder: (context) => AlertDialog(
                                      title: const Text('Load Project?'),
                                      content: const Text('Current workspace will be replaced. Make sure to save changes first.'),
                                      actions: [
                                        TextButton(
                                          onPressed: () => Navigator.pop(context),
                                          child: const Text('Cancel'),
                                        ),
                                        TextButton(
                                          onPressed: () {
                                            projectProvider.importFromJson(project.jsonContent);
                                            Navigator.pop(context); // Close dialog
                                            Navigator.pop(context); // Close library screen
                                            ScaffoldMessenger.of(context).showSnackBar(
                                              SnackBar(content: Text('Loaded project: ${project.name}')),
                                            );
                                          },
                                          child: const Text('Load'),
                                        ),
                                      ],
                                    ),
                                  );
                                },
                              ),
                              IconButton(
                                icon: const Icon(Icons.delete, color: Colors.red),
                                tooltip: 'Delete Project',
                                onPressed: () {
                                  showDialog(
                                    context: context,
                                    builder: (context) => AlertDialog(
                                      title: const Text('Delete Project?'),
                                      content: Text('Are you sure you want to delete "${project.name}"?'),
                                      actions: [
                                        TextButton(
                                          onPressed: () => Navigator.pop(context),
                                          child: const Text('Cancel'),
                                        ),
                                        TextButton(
                                          onPressed: () {
                                            libraryProvider.deleteProject(project.id);
                                            Navigator.pop(context);
                                          },
                                          child: const Text('Delete', style: TextStyle(color: Colors.red)),
                                        ),
                                      ],
                                    ),
                                  );
                                },
                              ),
                            ],
                          ),
                        ),
                      );
                    },
                  ),
          ),
        ],
      ),
    );
  }
}

File: src/screens/ProjectsLibraryScreen.js
// ProjectsLibraryScreen (React Native)
import React, { useMemo, useState } from 'react';
import { View, Text, TextInput, FlatList, StyleSheet, TouchableOpacity, Alert } from 'react-native';
import SaveToLibraryDialog from '../components/SaveToLibraryDialog';
import { useLibrary } from '../providers/LibraryProvider';
import { useProject } from '../context/ProjectContext';

const ProjectsLibraryScreen = () => {
  const { projects, deleteProject, saveProject } = useLibrary();
  const { exportToJson, importFromJson } = useProject();
  const [search, setSearch] = useState('');
  const [showSave, setShowSave] = useState(false);

  const filtered = useMemo(
    () => projects.filter((p) => p.name.toLowerCase().includes(search.toLowerCase())),
    [projects, search],
  );

  const handleLoad = (project) => {
    importFromJson(project.jsonContent);
    Alert.alert('Loaded', 'Project loaded into workspace.');
  };

  return (
    <View style={styles.container}>
      <Text style={styles.title}>Projects Library</Text>
      <View style={styles.actionsRow}>
        <TouchableOpacity style={styles.btn} onPress={() => setShowSave(true)}>
          <Text style={styles.btnText}>Save current project</Text>
        </TouchableOpacity>
      </View>
      <TextInput
        style={styles.input}
        placeholder="Search projects..."
        value={search}
        onChangeText={setSearch}
      />
      <FlatList
        data={filtered}
        keyExtractor={(item) => item.id}
        renderItem={({ item }) => (
          <View style={styles.card}>
            <View style={styles.cardHeader}>
              <Text style={styles.cardTitle}>{item.name}</Text>
              <TouchableOpacity onPress={() => deleteProject(item.id)}>
                <Text style={styles.deleteText}>Delete</Text>
              </TouchableOpacity>
            </View>
            <Text style={styles.cardSubtitle}>{item.description || 'No description'}</Text>
            {item.tags?.length ? <Text style={styles.tags}>Tags: {item.tags.join(', ')}</Text> : null}
            <View style={styles.cardActions}>
              <TouchableOpacity style={styles.smallBtn} onPress={() => handleLoad(item)}>
                <Text style={styles.smallBtnText}>Load</Text>
              </TouchableOpacity>
            </View>
          </View>
        )}
        ListEmptyComponent={<Text style={styles.empty}>No projects found.</Text>}
      />
      <SaveToLibraryDialog
        visible={showSave}
        onClose={() => setShowSave(false)}
        onSave={(payload) => {
          saveProject({ name: payload.name, description: payload.desc, tags: payload.tags || [], jsonContent: exportToJson() });
        }}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: '#F1F5F9', padding: 16 },
  title: { fontSize: 22, fontWeight: 'bold', marginBottom: 16 },
  actionsRow: { flexDirection: 'row', justifyContent: 'center', marginBottom: 12 },
  btn: { backgroundColor: '#6366F1', borderRadius: 8, paddingVertical: 10, paddingHorizontal: 16, marginHorizontal: 4 },
  btnText: { color: '#fff', fontWeight: 'bold' },
  input: { backgroundColor: '#fff', borderRadius: 8, padding: 8, marginBottom: 16, borderWidth: 1, borderColor: '#ddd' },
  card: { backgroundColor: '#fff', borderRadius: 12, padding: 16, marginBottom: 12, elevation: 2 },
  cardHeader: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center' },
  cardTitle: { fontSize: 16, fontWeight: 'bold' },
  cardSubtitle: { color: '#475569', marginTop: 4 },
  tags: { color: '#64748B', fontSize: 12, marginTop: 6 },
  cardActions: { flexDirection: 'row', marginTop: 12 },
  smallBtn: { backgroundColor: '#EEF2FF', borderRadius: 8, paddingVertical: 6, paddingHorizontal: 12 },
  smallBtnText: { color: '#4338CA', fontWeight: '700' },
  deleteText: { color: '#EF4444', fontWeight: '700', fontSize: 12 },
  empty: { color: '#64748B', textAlign: 'center', marginTop: 24 },
});

export default ProjectsLibraryScreen;
